### PR TITLE
metal: split multi-axis Reduce into single-axis reduces

### PR DESCRIPTION
**Pull request by czoli1976**
Ports the `split_multi_axis_reduce` rule from #2077 (cuda) to the Metal transform: a `Reduce` over several axes is rewritten into a chain of single-axis reduces before translation, so it stops falling back to CPU. `GpuReduce` only ever dispatches `axes[0]`, so a multi-axis reduce could not translate and the graph had to round-trip through the host in the middle.

`MeanOfSquares` is excluded, as in the cuda rule, since chaining it per axis is not the multi-axis result.

### Effect on MobileNet v2 (`--metal -O`, `dump --audit-json`)

The global average pool (`Reduce<Sum>` over two axes) was the only op left on host:

| | nodes | sync nodes | ops still on CPU |
|---|---|---|---|
| before | 376 | 4 | `Reduce<Sum>` |
| after | 375 | 2 | none |

The two remaining syncs are the model's own input and output.

Same seeded input through both binaries: max absolute difference `9.5e-07` over the 1000 logits, identical argmax.

Interleaved A/B, 5 rounds each, 300 loops per run, M-series:

```
before  13.119  12.629  11.616  12.842  13.183   best 11.616  mean 12.68
after   11.417  11.335  11.340  11.499  11.671   best 11.335  mean 11.45
```

Best-to-best that is 2.4%; the wider mean gap is mostly the run-to-run jitter the mid-graph flush was adding (before spans 1.57 ms, after 0.34 ms). The point of the change is removing the fallback rather than the timing.

### Checks

`cargo fmt --all` clean, `cargo clippy --workspace` clean, `tract-core` 268/268.

Two failures in `tract-gpu` / `tract-metal` reproduce identically on `main` with this commit stashed out, so they are not from this change: `tensor::tests::test_device_tensor` ("GPU Context not initialized") and `kernels::matmul::mfa::tests::test_mfa_attention_causal_const_is_noop` (the causal probe returns `full=0.08356` where it expects the triangular-const-alone value). Happy to open a separate issue for the latter if it is not already known.

🍍

### CODE REVIEW

**Comment by kali** [comment]
> Two failures in `tract-gpu` / `tract-metal` reproduce identically on `main` with this commit stashed out, so they are not from this change: `tensor::tests::test_device_tensor` ("GPU Context not initialized") and `kernels::matmul::mfa::tests::test_mfa_attention_causal_const_is_noop` (the causal probe returns `full=0.08356` where it expects the triangular-const-alone value). Happy to open a separate issue for the latter if it is not already known.

I'm not aware of these. Do we have a gap in the CI coverage ?

File: metal/src/transform.rs [hunk 1]
```diff
@@ -503,3 +505,28 @@ fn convert_const(op: &Const) -> TractResult<Const> {
     let metal_const = op.val().clone().into_device()?.into_tensor().into_arc_tensor();
     Const::new_with_exotic_fact(metal_const, Box::new(metal_fact))
 }
+
+/// Rewrites a `Reduce` over several axes into a chain of single-axis reduces, which is
+/// what `GpuReduce` accepts. Only reducers that compose associatively per axis qualify:
+/// `MeanOfSquares` over a chain is not the multi-axis result.
+fn split_multi_axis_reduce(
```
Comment (kali) [inline metal/src/transform.rs:512]: is this a function that already exists in cuda ? (just guessing from the commit message). Can we unify them in tract-gpu ?

Comment (czoli1976) [reply metal/src/transform.rs:512]: Yes — it's the rule from #2077; I mirrored it rather than sharing it because `add_matmul_broadcast` / `fuse_axis_op` / `untranspose_matmul_output` are all per-backend duplicates today. Unified now: it lives in `tract_gpu::rewrite_rules::split_multi_axis_reduce` and both transforms import it, so the cuda copy goes away too and the PR is net -24 lines. Same graph and same outputs on MobileNet as before the hoist.

**Comment by czoli1976** [comment]
On the two failures — they turn out to be different stories.

The MFA one is not a CI gap: `crates.yml`'s `metal` job runs `cargo test -p tract-metal -p test-metal` on every PR, and `metal (1.91)` passed on this PR's commit. So it passes on your runner and fails only on mine. This machine is on a macOS beta, so my guess is a driver/OS difference in the causal path rather than a code bug — happy to dig into it separately and open an issue if it turns out to be real.

`tract-gpu` does look like a genuine gap though: nothing in `.github/workflows` references it, so `cargo test -p tract-gpu` never runs anywhere. Standalone it cannot pass — `tensor::tests::test_device_tensor` fails on "GPU Context not initialized" because no backend has registered a context in that test binary.

Unrelated, for the record: the bench bot's 4 secondary regressions are all `arm_ml_kws_cnn_m` load / load+optimize on cortex-a7 and cortex-a9, where `tract-metal` isn't even a dependency (`cli/Cargo.toml` gates it behind `cfg(any(target_os = "macos", target_os = "ios"))`), so they can't come from this change.

File: metal/src/transform.rs (hunk 1 above)
Comment (czoli1976) [reply metal/src/transform.rs:512]: moved to tract-gpu

**Comment by kali** [comment]
About the load, I tried to load in two times to split away the i/o from the actual loading. We'll see. I acknowledge it's borderline useless today.

**Comment by kali** [comment]
@czoli1976 i'm going to remove my two commits from there. it was opportunistic, and it failed spectacularly. Moving them to a different PR

**Comment by kali** [comment]
⚠️⚠️⚠️ Just rebased! ⚠️⚠️⚠️
